packages/asset-buyer/src/quote_consumers/forwarder_swap_quote_consumer.ts

```diff
@@ -0,0 +1,228 @@
+        }
+        assert.isETHAddressHex('feeRecipient', feeRecipient);
+        if (gasLimit !== undefined) {
+            assert.isNumber('gasLimit', gasLimit);
+        }
+        if (gasPrice !== undefined) {
+            assert.isBigNumber('gasPrice', gasPrice);
+        }
+
+        const swapQuoteWithFeeAdded = addAffiliateFeeToSwapQuote(quote, feePercentage);
+
+        const { orders, feeOrders, makerAssetBuyAmount, worstCaseQuoteInfo } = swapQuoteWithFeeAdded;
+
+        const finalTakerAddress = await this._getTakerAddressOrThrowAsync(opts);
+        try {
+            const txHash = await this._contractWrappers.forwarder.marketBuyOrdersWithEthAsync(
+                orders,
+                makerAssetBuyAmount,
+                finalTakerAddress,
+                ethAmount || worstCaseQuoteInfo.totalTakerTokenAmount,
+                feeOrders,
+                feePercentage,
+                feeRecipient,
+                {
+                    gasLimit,
+                    gasPrice,
+                    shouldValidate: true,
+                },
+            );
+            return txHash;
+        } catch (err) {
+            if (_.includes(err.message, ContractWrappersError.SignatureRequestDenied)) {
+                throw new Error(AssetSwapQuoterError.SignatureRequestDenied);
+            } else if (_.includes(err.message, ForwarderWrapperError.CompleteFillFailed)) {
+                throw new Error(AssetSwapQuoterError.TransactionValueTooLow);
+            } else {
+                throw err;
+            }
+        }
+    }
+
+    public isValidForwarderSwapQuote = (quote: SwapQuote): boolean => {
+        const isValidMarketOrders = this._isValidForwarderSignedOrders(quote.orders);
+        const isValidFeeOrders = this._isValidForwarderSignedOrders(quote.feeOrders);
+        return isValidMarketOrders && isValidFeeOrders;
+    }
+
+    private readonly _getTakerAddressOrThrowAsync = async (opts: Partial<ForwarderSwapQuoteExecutionOpts>): Promise<string> => {
+        if (opts.takerAddress !== undefined) {
+            return opts.takerAddress;
+        } else {
+            const web3Wrapper = new Web3Wrapper(this.provider);
+            const availableAddresses = await web3Wrapper.getAvailableAddressesAsync();
+            const firstAvailableAddress = _.head(availableAddresses);
+            if (firstAvailableAddress !== undefined) {
+                return firstAvailableAddress;
+            } else {
+                throw new Error(AssetSwapQuoterError.NoAddressAvailable);
+            }
+        }
+    }
+
+    private _getEtherTokenAssetDataOrThrow(): string {
+        return assetDataUtils.getEtherTokenAssetData(this._contractWrappers);
+    }
+
+    private readonly _isValidForwarderSignedOrders = (orders: SignedOrder[]): boolean => {
+        const wethAssetData = this._getEtherTokenAssetDataOrThrow();
+        return _.reduce(orders, (a: boolean, c: SignedOrder): boolean => {
```

Comment:
I feel like a `.filter` + length would be easier to understand here. Or just `every` https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/every